### DISCUSSION TITLE
日本語対応の超高速全文検索機能

### DISCUSSION THREAD

**Discussion by cuongph1** [pgroonga/pgroonga#173] [Q&A] [answered]
PGroongaホムページに”PGroongaをPostgreSQLにインストールすると全言語対応の超高速全文検索機能を使えるようになります”を記載していますがPGroongaを導入してSQL実行によるパフォーマンスの向上の割合を調査したら結果は変わらない見えました

検証環境：Windows10
インストール：https://pgroonga.github.io/ja/install/windows.html
ガイド：https://pgroonga.github.io/ja/tutorial/
上記で検証を行っています

PGAdmin4ツールでtestテーブル作成＆インデックス作成
```
create table test (	
    id int,
    name VARCHAR ( 500 ) ,
    contents text
)
CREATE INDEX testindex
          ON test 
       USING pgroonga (name);
```
```
insert into test (id, name, contents) values (1, '文字列テストデータ文字列テストデータ',  '64bitバージョンのWindowsでも構いません。PostgreSQLがサポートしているすべてのWindowsで使えます。');
```
上記で1010000レコードを登録する

SQLの実行時間を計測する
```
SET enable_seqscan = off;
explain analyze select * from test where name &@ '文字列'
```
結果：

　　　　PGroonga導入しない純正Posgres環境｜純正PosgresにPGRoonga導入済み
1回目　　Execution Time: 4547.992 ms　　　  | Execution Time: 4593.014 ms
2回目　　Execution Time: 4538.367 ms　　　  | Execution Time: 4526.292 ms
3回目　　Execution Time: 4551.799 ms　　　  | Execution Time: 4540.022 ms

SQLの実行時間が変わらないのを見えます。上記で検証手順は間違ったところがあるかもしれませんが解決できないので経験がある方が教えていただけると助かります

**Comment by komainu8** [comment]
以下のSQLの結果を教えていただけますでしょうか？

```sql
EXPLAIN ANALYZE VERBOSE SELECT * FROM test WHERE name &@ '文字列';
```

あと、こちらでも再現させたいので、実際にデータを挿入した際のSQL(INSERT文)を提供いただけますでしょうか？

**Comment by komainu8** [comment, marked as answer]
あぁ。。。すみません。
原因がわかりました。

`name`が`varchar`型なので、今のインデックス定義では、PGroongaのインデックスを使って全文検索をしません。
PGroongaのインデックスが使われないので、PGroonga導入前と同じくらいの速度になっています。

`name`を`text`型にするか、`varchar`型のカラムに対して全文検索するならインデックスに`pgroonga_varchar_full_text_search_ops_v2`を指定する必要があります。

つまり、

```sql
create table test (	
    id int,
    name text ,
    contents text
)
CREATE INDEX testindex
          ON test 
       USING pgroonga (name);
```

あるいは

```sql
create table test (	
    id int,
    name VARCHAR ( 500 ) ,
    contents text
)
CREATE INDEX testindex
          ON test 
       USING pgroonga (name pgroonga_varchar_full_text_search_ops_v2);
```

とすれば、PGroongaのインデックスを使用して全文検索するので、高速に検索結果が帰ってきます。

**Reply by cuongph1** [reply]
返事ありがとうございます。
testindexを削除して、データが登録された状態で下記に再度インデックス作成しても問題ないでしょうか
~~~
CREATE INDEX testindex
          ON test 
       USING pgroonga (name pgroonga_varchar_full_text_search_ops_v2);
~~~

**Reply by komainu8** [reply]
はい。問題ありません。
以下のように実行いただければと思います。

```sql
DROP INDEX testindex;
CREATE INDEX testindex
          ON test 
       USING pgroonga (name pgroonga_varchar_full_text_search_ops_v2);
```

**Reply by cuongph1** [reply]
testindexを削除して、データが登録された状態で下記に再度インデックス作成しました

```
CREATE INDEX testindex
          ON test 
       USING pgroonga (name pgroonga_varchar_full_text_search_ops_v2);
```

次の検証は結果がまだ変わらないですが
```
EXPLAIN ANALYZE VERBOSE SELECT * FROM test WHERE name &@ '文字列';
```
を実行して下記結果になります（ヒットレコードが1100000、大体5~秒になります）
```
Index Scan using testindex on j0000001_ec.test  (cost=0.00..128824.01 rows=1100 width=552) (actual time=294.797..3400.494 rows=1100000 loops=1)"
  Output: id, name, contents"
  Index Cond: (test.name &@ '文字列'::character varying)"
Planning Time: 0.569 ms"
Execution Time: 5733.466 ms"
```

ちなみに
contentsのtext型もインデックス作成してもcontents条件として検索してみたら同じ事象になります
```
CREATE INDEX testindex_contents
          ON test 
       USING pgroonga (contents);
```
```
EXPLAIN ANALYZE VERBOSE SELECT * FROM test WHERE contents &@ 'bit';
```

```
Index Scan using testindex_contents on j0000001_ec.test  (cost=0.00..127724.00 rows=1100 width=552) (actual time=165.665..3277.634 rows=1100000 loops=1)"
  Output: id, name, contents"
  Index Cond: (test.contents &@ 'bit'::text)"
Planning Time: 10.605 ms"
Execution Time: 5609.907 ms"
```

**Reply by kou** [reply]
検索ではなく出力に時間がかかっているだけだったりしませんか？
`SELECT COUNT(*) ...`にしても同じ結果になりますか？

**Reply by cuongph1** [reply]
報告させていただきます。
①
```
EXPLAIN ANALYZE VERBOSE SELECT count(*) FROM test WHERE name &@ '文字列';
```
の結果
```
Aggregate  (cost=128826.76..128826.77 rows=1 width=8) (actual time=5839.157..5839.165 rows=1 loops=1)"
  Output: count(*)"
  ->  Index Scan using testindex on j0000001_ec.test  (cost=0.00..128824.01 rows=1100 width=0) (actual time=297.071..3414.385 rows=1100000 loops=1)"
        Output: id, name, contents"
        Index Cond: (test.name &@ '文字列'::character varying)"
Planning Time: 0.406 ms"
Execution Time: 5845.041 ms"
```

②
```
SELECT count(*) FROM test WHERE name &@ '文字列';
```
の結果
![ex1](https://user-images.githubusercontent.com/79302278/108481447-9057a780-72db-11eb-8680-8bf1c9061e16.JPG)

確かに表示まで時間が違いますね

**Reply by cuongph1** [reply]
申し訳ございませんが
上記形で検証してPGroonga導入前と時間が改善していないようです
PGAdmin4でSQLを実行し結果画面をキャプチャーを報告致します。

①PGroonga導入済みの場合

![ex2](https://user-images.githubusercontent.com/79302278/108482894-6acb9d80-72dd-11eb-9737-7a5c2e3a7faa.JPG)

①PGroonga導入前の場合
![ex3](https://user-images.githubusercontent.com/79302278/108482933-74ed9c00-72dd-11eb-9030-10d686c25b5c.JPG)
※PGroonga導入済み環境からDROP EXTENSION pgroonga CASCADE;を実行して検証しました

**Reply by kou** [reply]
すべてのレコードがヒットするクエリーを試しているように見えるのですが、すべてのレコードがヒットするクエリーを速くしたいのですか？
であれば、`WHERE`を指定せずに`SELECT * FROM test`とするのが最速だと思います。

**Reply by cuongph1** [reply]
そうじゃなくて、`SELECT * FROM test`の方が最速となるのは当然だと思いますが
これは簡単なテーストデータなので全てのレコードが同じになってて全文検索の結果がすべてのレコードがヒットされしまいました。
実際のプロジェクトがいろいろなケースがあります。
希望はPGroonga導入して、`&@ '文字列'`みたいな条件で検証するのが私の場合に対して本当に高速化になるかどうかですね。

**Reply by kou** [reply]
実際に検索したいデータがある場合は、実際のデータで性能を検証してください。
たとえテストデータで速くても実際のデータで遅ければ意味がありません。
あるいは、逆に、テストデータで遅くても実際のデータで速ければ有用です。

**Reply by cuongph1** [reply]
全てのレコードをヒットする場合、検索の速度が出ないのですが一部ヒットの場合検索の速度が速くなることを確認できました。

**Reply by kou** [reply]
よかったです！
